validator/machine_loader.go

```diff
@@ -0,0 +1,184 @@
+package validator
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type JitMachineConfig struct {
+	ProverBinPath string
+	JitCranelift  bool
+}
+
+var DefaultJitMachineConfig = JitMachineConfig{
+	JitCranelift:  true,
+	ProverBinPath: "replay.wasm",
+}
+
+type machineStatus[M any] struct {
+	machine   *M
+	readyChan chan struct{}
+	err       error
+}
+
+func newMachineStatus[M any]() *machineStatus[M] {
+	return &machineStatus[M]{
+		readyChan: make(chan struct{}),
+	}
+}
+
+type machineLoader[M any] struct {
+	mapMutex            sync.Mutex
+	machines            map[common.Hash]*machineStatus[M]
+	locator             *MachineLocator
+	createMachineThread func(ctx context.Context, moduleRoot common.Hash, status *machineStatus[M])
+	stopped             bool
+}
+
+func getJitPath() (string, error) {
+	var jitBinary string
+	executable, err := os.Executable()
+	if err == nil {
+		if strings.Contains(filepath.Base(executable), "test") {
+			_, thisfile, _, _ := runtime.Caller(0)
+			projectDir := filepath.Dir(filepath.Dir(thisfile))
+			jitBinary = filepath.Join(projectDir, "target", "bin", "jit")
+		} else {
+			jitBinary = filepath.Join(filepath.Dir(executable), "jit")
+		}
+		_, err = os.Stat(jitBinary)
+	}
+	if err != nil {
+		var lookPathErr error
+		jitBinary, lookPathErr = exec.LookPath("jit")
+		if lookPathErr == nil {
+			return jitBinary, nil
+		}
+	}
+	return jitBinary, err
+}
+
+func newMachineLoader[M any](locator *MachineLocator,
+	createMachineThread func(ctx context.Context, moduleRoot common.Hash, status *machineStatus[M])) *machineLoader[M] {
```

Comment:
It doesn't really matter but I personally find the formatting
```go
func foo(
  arg1 int,
  arg2 int,
) int {
}
```
a bit cleaner if it needs split across multiple lines